### ISSUE TITLE
2D context implementation completeness chart

### ISSUE THREAD

**Issue by tailsu** [NativeScript/nativescript-canvas#1]
Pull requests implementing this or that are welcome.

# iOS

## Properties
- [ ] canvas
- [ ] fillStyle
- [ ] font
- [ ] globalAlpha
- [ ] globalCompositeOperation
- [ ] lineCap
- [ ] lineDashOffset
- [ ] lineJoin
- [ ] lineWidth
- [ ] miterLimit
- [ ] shadowBlur
- [ ] shadowColor
- [ ] shadowOffsetX
- [ ] shadowOffsetY
- [ ] strokeStyle
- [ ] textAlign
- [ ] textBaseline

## Methods
- [ ] arc
- [ ] arcTo
- [ ] beginPath
- [ ] bezierCurveTo
- [ ] clearRect
- [ ] clip
- [ ] closePath
- [ ] createImageData
- [ ] createLinearGradient
- [ ] createPattern
- [ ] createRadialGradient
- [ ] drawFocusIfNeeded
- [ ] drawImage
- [ ] fill
- [ ] fillRect
- [ ] getImageData
- [ ] getLineDash
- [ ] isPointInPath
- [ ] isPointInStroke
- [ ] lineTo
- [ ] measureText
- [ ] moveTo
- [ ] putImageData
- [ ] quadraticCurveTo
- [ ] rect
- [ ] restore
- [ ] rotate
- [ ] save
- [ ] scale
- [ ] setLineDash
- [ ] setTransform
- [ ] stroke
- [ ] strokeRect
- [ ] strokeText
- [ ] transform
- [ ] translate

# Android

## Properties
- [ ] canvas
- [ ] fillStyle
- [ ] font
- [ ] globalAlpha
- [ ] globalCompositeOperation
- [ ] lineCap
- [ ] lineDashOffset
- [ ] lineJoin
- [ ] lineWidth
- [ ] miterLimit
- [ ] shadowBlur
- [ ] shadowColor
- [ ] shadowOffsetX
- [ ] shadowOffsetY
- [ ] strokeStyle
- [ ] textAlign
- [ ] textBaseline

## Methods
- [ ] arc
- [ ] arcTo
- [ ] beginPath
- [ ] bezierCurveTo
- [ ] clearRect
- [ ] clip
- [ ] closePath
- [ ] createImageData
- [ ] createLinearGradient
- [ ] createPattern
- [ ] createRadialGradient
- [ ] drawFocusIfNeeded
- [ ] drawImage
- [ ] fill
- [ ] fillRect
- [ ] getImageData
- [ ] getLineDash
- [ ] isPointInPath
- [ ] isPointInStroke
- [ ] lineTo
- [ ] measureText
- [ ] moveTo
- [ ] putImageData
- [ ] quadraticCurveTo
- [ ] rect
- [ ] restore
- [ ] rotate
- [ ] save
- [ ] scale
- [ ] setLineDash
- [ ] setTransform
- [ ] stroke
- [ ] strokeRect
- [ ] strokeText
- [ ] transform
- [ ] translate

**Comment by harangue** [comment]
If this comes to completion, it would be BEYOND cool. Wish I had the native mobile dev experience to help out!

**Comment by icode** [comment]
Did not continue the work?

**Comment by tailsu** [comment]
**TL;DR:** NativeScript is not good for iOS games.

Nope. I'd say that there's no point to, specifically for iOS. Here's why.

Immediate-mode drawing API's like Canvas benefit from fast execution - drawing calls usually do something very small and there are a lot of drawing calls per frame. Having little overhead per call is a significant benefit.

We're talking JavaScript, so what we have to compare here is:

1. An app with an embedded JavaScriptCore (like NS) plus bindings to the iOS drawing API
2. The iOS WebView running an HTML5 Canvas

In the context of immediate-mode API's here is the clincher - JIT.

As long as iOS apps are not allowed to have JIT compilation, JavaScriptCore will remain an interpreted runtime. The iOS WebView is hosted in a blessed process that allows JIT'ting, which essentially means, that **apps leveraging HTML/Canvas will be faster than a native app leveraging canvas bindings**.

The argument is moot if you don't care about performance and the usage of the native UI controls overrules all other arguments. However, regarding the Canvas API, it is rare for an app to use it as a satellite feature. It is more likely to be structured completely around it, like, for example, a game. In general, if you use an immediate-mode API, you care about performance.

So, on one hand you have HTML5/Canvas/JavaScript which is widely supported and has a bunch of engines and libraries ready for use. On the other hand you have NativeScript with tenuous Canvas support and a slow JS VM to boot.

If you're looking for *performance* in a portable (hence canvas-based), heavily animated app, then HTML is your best bet for iOS (Android doesn't have the JIT limitation, so performance characteristics on it depend on completely different arguments).

PS: Correct me if I've become wrong about iOS and JIT compilation.

**Comment by icode** [comment]
this for js(canvas) chart is very useful, not need fast
